### PR TITLE
UNDERTOW-898 Failover targets should be chosen deterministically

### PR DESCRIPTION
**Pull request by rhusar**
Jira
https://issues.jboss.org/browse/UNDERTOW-898

### CODE REVIEW

**Comment by undertow-pull-request** [comment]
**Linux** [Build 221](https://ci.wildfly.org/viewLog.html?buildId=37317&buildTypeId=UNDERTOW_14xLinux) outcome was **FAILURE** using a merge of 92198ccebdc7dd56f45a32717b21e4d3039fa4fa
Summary: Tests failed: 1 (1 new), passed: 1721, ignored: 449 Build time: 00:10:23
### Failed tests
```
io.undertow.server.handlers.LotsOfQueryParametersTestCase.testLotsOfQueryParameters_MaxParameters_Ok{proxy}{https}: java.lang.RuntimeException: Buffer Leak io.undertow.server.handlers.LotsOfQueryParametersTestCase.testLotsOfQueryParameters_MaxParameters_Ok{proxy}{https}
Caused by: java.lang.RuntimeException: java.nio.DirectByteBuffer[pos=0 lim=17408 cap=17408] NO: 181381 [NO_CONTEXT]
------- Stderr: -------
java.lang.RuntimeException: java.nio.DirectByteBuffer[pos=0 lim=17408 cap=17408] NO: 181381 [NO_CONTEXT]
	at io.undertow.testutils.DebuggingSlicePool$DebuggingBuffer.<init>(DebuggingSlicePool.java:84)
	at io.undertow.testutils.DebuggingSlicePool.allocate(DebuggingSlicePool.java:45)
	at io.undertow.protocols.ssl.SslConduit.doUnwrap(SslConduit.java:696)
	at io.undertow.protocols.ssl.SslConduit.doHandshake(SslConduit.java:645)
	at io.undertow.protocols.ssl.SslConduit.access$900(SslConduit.java:63)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.readReady(SslConduit.java:1097)
	at org.xnio.nio.NioSocketConduit.handleReady(NioSocketConduit.java:88)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:559)


```

File: core/src/main/java/io/undertow/server/handlers/proxy/mod_cluster/ModClusterContainer.java [hunk 1]
```diff
@@ -394,10 +395,22 @@ Context findNewNode(final VirtualHost.HostEntry entry) {
...
      * @param jvmRoute the original jvmRoute
+     * @param entry    the resolved virtual host entry
      * @return the context, {@code null} if not found
-     * @oaram entry      the resolved virtual host entry
      */
-    Context findFailoverNode(final VirtualHost.HostEntry entry, final String domain, final String jvmRoute, final boolean forceStickySession) {
+    Context findFailoverNode(final VirtualHost.HostEntry entry, final String domain, final String session, final String jvmRoute, final boolean forceStickySession) {
+
+        // If configured, deterministically choose the failover target by calculating hash of the session ID modding by number of electable nodes
+        if (modCluster.isDeterministicFailover()) {
```
Comment (stuartwdouglas) [inline core/src/main/java/io/undertow/server/handlers/proxy/mod_cluster/ModClusterContainer.java]: Is this guaranteed to always get a live node? Or can it pick nodes in error state?

**Comment by undertow-pull-request** [comment]
**Linux** [Build 226](https://ci.wildfly.org/viewLog.html?buildId=37444&buildTypeId=UNDERTOW_14xLinux) outcome was **FAILURE** using a merge of 2ed926445b2bd694f7fd0da66909455e110bfa3c
Summary: Tests passed: 1722, ignored: 449; exit code 1 Build time: 00:10:22

**Comment by undertow-pull-request** [comment]
**Linux** [Build 227](https://ci.wildfly.org/viewLog.html?buildId=37445&buildTypeId=UNDERTOW_14xLinux) outcome was **FAILURE** using a merge of 2ed926445b2bd694f7fd0da66909455e110bfa3c
Summary: Tests failed: 1 (1 new), passed: 1721, ignored: 449 Build time: 00:09:56
### Failed tests
```
io.undertow.server.handlers.LotsOfHeadersRequestTestCase.testLotsOfHeadersInRequest_MaxHeaders_BadRequest{proxy}{https}: java.lang.RuntimeException: Buffer Leak io.undertow.server.handlers.LotsOfHeadersRequestTestCase.testLotsOfHeadersInRequest_MaxHeaders_BadRequest{proxy}{https}
Caused by: java.lang.RuntimeException: java.nio.DirectByteBuffer[pos=0 lim=17408 cap=17408] NO: 178770 [NO_CONTEXT]
------- Stderr: -------
java.lang.RuntimeException: java.nio.DirectByteBuffer[pos=0 lim=17408 cap=17408] NO: 178770 [NO_CONTEXT]
	at io.undertow.testutils.DebuggingSlicePool$DebuggingBuffer.<init>(DebuggingSlicePool.java:84)
	at io.undertow.testutils.DebuggingSlicePool.allocate(DebuggingSlicePool.java:45)
	at io.undertow.protocols.ssl.SslConduit.doUnwrap(SslConduit.java:696)
	at io.undertow.protocols.ssl.SslConduit.doHandshake(SslConduit.java:645)
	at io.undertow.protocols.ssl.SslConduit.access$900(SslConduit.java:63)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.readReady(SslConduit.java:1097)
	at org.xnio.nio.NioSocketConduit.handleReady(NioSocketConduit.java:88)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:559)


```

**Comment by undertow-pull-request** [comment]
**Linux** [Build 228](https://ci.wildfly.org/viewLog.html?buildId=37446&buildTypeId=UNDERTOW_14xLinux) outcome was **FAILURE** using a merge of 2ed926445b2bd694f7fd0da66909455e110bfa3c
Summary: Tests passed: 1722, ignored: 449; exit code 1 Build time: 00:09:54

**Comment by rhusar** [comment]
Closing for now, will reopen with 2 more fixes.

**Comment by rhusar** [comment]
Superseeded by https://github.com/undertow-io/undertow/pull/460 as it includes another necessari fix.

File: core/src/main/java/io/undertow/server/handlers/proxy/mod_cluster/ModClusterContainer.java (hunk 1 above)
Comment (rhusar) [inline core/src/main/java/io/undertow/server/handlers/proxy/mod_cluster/ModClusterContainer.java]: By broken feature i meant hot standby feature in modcluster, see https://issues.jboss.org/browse/UNDERTOW-932

**Pull request closed without merging**